### PR TITLE
fix(prometheus): conflict between global rule and route configure

### LINKED ISSUE #6555: bug:  Prometheus metrics igress and egress is inaccurate
### Issue description

There two routes, one priority is 0 route id = 1, anther is 30 route id = 3.

if request hit route 1, the prometheus metric igress and egress is accurate, but if request hit route 3, the igress and egress valuse is double of bandwidth.

I think that apisix records twice `request_length` and `bytes_sent`.

```
    metrics.bandwidth:inc(vars.request_length,
        gen_arr("ingress", route_id, service_id, consumer_name, balancer_ip))

    metrics.bandwidth:inc(vars.bytes_sent,
        gen_arr("egress", route_id, service_id, consumer_name, balancer_ip))
```

### Environment

- apisix version (cmd: `apisix version`): 2.7
- OS (cmd: `uname -a`):
- OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):
- etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API):
- apisix-dashboard version, if have:
- the plugin runner version, if the issue is about a plugin runner (cmd: depended on the kind of runner):
- luarocks version, if the issue is about installation (cmd: `luarocks --version`):

### Steps to reproduce

route configuration
```
{
  "action": "get",
  "node": {
    "key": "/apisix/routes",
    "dir": true,
    "nodes": [
      {
        "key": "/apisix/routes/1",
        "modifiedIndex": 6332825,
        "value": {
          "status": 1,
          "methods": [
            "PUT",
            "GET",
            "POST",
            "DELETE",
            "PATCH",
            "HEAD",
            "OPTIONS",
            "CONNECT",
            "TRACE"
          ],
          "upstream": {
            "pass_host": "pass",
            "nodes": {
              "10.235.79.7:8079": 10
            },
            "scheme": "http",
            "hash_on": "vars",
            "type": "roundrobin"
          },
          "uri": "/*",
          "create_time": 1646098895,
          "id": "1",
          "priority": 0,
          "update_time": 1646098895,
          "plugins": {
            "prometheus": {
              "prefer_name": false
            }
          }
        },
        "createdIndex": 6332825
      },
      {
        "key": "/apisix/routes/3",
        "modifiedIndex": 4020996,
        "value": {
          "status": 1,
          "methods": [
            "PUT",
            "GET",
            "POST",
            "DELETE",
            "PATCH",
            "HEAD",
            "OPTIONS",
            "CONNECT",
            "TRACE"
          ],
          "upstream": {
            "hash_on": "vars",
            "type": "roundrobin",
            "nodes": {
              "10.235.79.7:9096": 10
            },
            "pass_host": "pass",
            "scheme": "http"
          },
          "uri": "/*",
          "priority": 20,
          "vars": [
            "OR",
            [
              "host",
              "~*",
              ".*(abc.cn)$"
            ],
            [
              "host",
              "==",
              "10.235.82.1"
            ]
          ],
          "id": "3",
          "update_time": 1643262492,
          "create_time": 1638353491
        },
        "createdIndex": 1682
      }
    ]
  }
}

```
Client write: 145MB/s + 133MB/s + 144MB/s + 1.15GB/s = 1.55GB/s, but apisix igress is 2.85GB/s

image

image

### Actual result

The bandwidth = apisix igress or egress

### Error log

no

### Expected result

_No response_

### PR DESCRIPTION
**Pull request by shuaijinchao**
### What this PR does / why we need it:
FIX #6555

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/plugins/prometheus.lua [hunk 1]
```diff
@@ -39,6 +39,7 @@ local _M = {
     name = plugin_name,
     log  = exporter.log,
     schema = schema,
+    run_policy = "prefer_route",
```
Comment (zhendongcmss) [inline apisix/plugins/prometheus.lua:42]: Does the `nginx-lua-prometheus` need update ?

Comment (zhendongcmss) [reply apisix/plugins/prometheus.lua:42]: This fix like https://github.com/apache/apisix/pull/4589/files

Comment (shuaijinchao) [reply apisix/plugins/prometheus.lua:42]: > Does the `nginx-lua-prometheus` need update ?

No, APISIX from 2.7 to master, `nginx-lua-prometheus` has only one version upgrade, and this upgrade of `nginx-lua-prometheus` is not related to the current issue.

https://github.com/knyar/nginx-lua-prometheus/blob/master/CHANGELOG.md#020220127

**Review by tzssangglass — APPROVED** [review]

**Review by tokers — DISMISSED** [review]
The way do work for fixing the problem but my concern is that are there other plugins have the same problem? Shall we have a generic mechanism to handle this situation?

File: apisix/plugins/prometheus/exporter.lua [hunk 2]
```diff
@@ -171,6 +171,8 @@ function _M.log(conf, ctx)
 
     metrics.bandwidth:inc(vars.bytes_sent,
         gen_arr("egress", route_id, service_id, consumer_name, balancer_ip))
+
+    core.log.info("prometheus run report")
```
Comment (spacewander) [inline apisix/plugins/prometheus/exporter.lua:175]: Can we check the result of prometheus instead of adding a new log just for testing?

**Comment by spacewander** [comment]
> The way do work for fixing the problem but my concern is that are there other plugins have the same problem? Shall we have a generic mechanism to handle this situation?

This has been discussed before. The `run_policy` used in this PR is the solution to it.

File: apisix/plugins/prometheus/exporter.lua (hunk 2 above)
Comment (shuaijinchao) [reply apisix/plugins/prometheus/exporter.lua:175]: this is not just for testing purposes. Currently, there is no log output when prometheus reports data. I think adding this log will also help developers watch it while developing and debugging.

what do you think?

Comment (spacewander) [reply apisix/plugins/prometheus/exporter.lua:175]: A hit will increase the metric: https://github.com/apache/apisix/blob/1ac7166e80fdd502cc6a8a60ef58f415d3c0e56c/apisix/plugins/prometheus/exporter.lua#L153

There is no need to add another way.

File: t/plugin/prometheus3.t [hunk 3]
```diff
@@ -256,7 +256,12 @@ passed
 GET /opentracing
 --- response_body
 opentracing
---- grep_error_log eval
-qr/prometheus run \w+/
---- grep_error_log_out
-prometheus run report
+--- error_code: 200
```
Comment (spacewander) [inline t/plugin/prometheus3.t:259]: The error_code check here is redundant?

File: t/plugin/prometheus3.t [hunk 4]
```diff
@@ -256,7 +256,12 @@ passed
...
+
+
+
+=== TEST 7: fetch prometheus plugin at both global rule and route data
+--- request
+GET /apisix/prometheus/metrics
+--- response_body eval
+qr/apisix_http_status\{code="200",route="1",matched_uri="\/opentracing",matched_host="",service="",consumer="",node="127.0.0.1\"} \d+/
```
Comment (spacewander) [inline t/plugin/prometheus3.t]: It doesn't check the number this request is hit...

**Review by spacewander — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**